integrations/terraform-cloud/client.py

```diff
@@ -0,0 +1,286 @@
+
+                for workspace in workspaces:
+                    workspace_name = workspace['attributes']['name']
+                    organization_id = workspace['relationships']['data']['id']
+                    page = 1
+
+                    all_state_versions = []
+                    while True:              
+                        
+                        state_versions_url = f"{self.api_url}/state-versions?page[number]={page}&page[size]={PAGE_SIZE}&filter[workspace][name]={workspace_name}&filter[organization][name]={organization_id}"
+                        response = await self.client.get(state_versions_url)
+                        response.raise_for_status()
+                        state_versions_data = response.json()
+                        state_versions = state_versions_data.get('data', [])
+                        yield state_versions
+
+                        if len(state_versions) < PAGE_SIZE:
+                            break
+
+                        page += 1
+
+                        all_state_versions.extend(workspaces)
+
+                    event.attributes[cache_key] = all_state_versions
+
+        except httpx.HTTPStatusError as e:
+            logger.error(f"HTTP error for state versions: Status {e.response.status_code}, Response: {e.response.text}")
+        except httpx.HTTPError as e:
+            logger.error(f"Error fetching terraform data: {e}")
+
+
+
+    async def get_organizations(self)-> List[Dict[str, Any], None]:
+        
+        try:
+            logger.info("Getting organizations from terraform")
+
+            organizations_url = f"{self.api_url}/organizations"
+            response = await self.client.get(organizations_url)
+            response.raise_for_status()
+            organizations_data = response.json()
+            organizations = organizations_data.get('data', [])
+            return organizations
+
+        
+        except httpx.HTTPStatusError as e:
+            logger.error(f"HTTP error for state versions: Status {e.response.status_code}, Response: {e.response.text}")
+        except httpx.HTTPError as e:
+            logger.error(f"Error fetching terraform data: {e}")
+
+        
+
+    async def get_state_version_outputs_for_workspace(self)-> AsyncGenerator[Dict[str, Any], None]:
+        try:
+
+            page = 1
+            while True:
+
+                async for state_versions in self.get_state_version_for_workspace():
+                    for state_version in state_versions:
+                        logger.info("Getting state version outputs for all workspaces")
+                        state_version_outputs_url = f"{self.api_url}/state-versions/{state_version['id']}/outputs?page[number]={page}&page[size]={PAGE_SIZE}"
+                        response = await self.client.get(state_version_outputs_url)
+                        response.raise_for_status()
+                        state_version_outputs_data = response.json()

```

Comment:
Since almost all the response gives data object, you could create a single function that makes the rest api call. That way, we wouldn't have to  repeat this section across in the code. The function could take the url, and the request params